core/trino-main/src/main/java/io/trino/sql/planner/optimizations/DetermineHashPartitionCount.java

```diff
@@ -0,0 +1,192 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.optimizations;
+
+import io.airlift.log.Logger;
+import io.trino.Session;
+import io.trino.cost.CachingStatsProvider;
+import io.trino.cost.StatsCalculator;
+import io.trino.cost.StatsProvider;
+import io.trino.cost.TableStatsProvider;
+import io.trino.execution.warnings.WarningCollector;
+import io.trino.sql.planner.PlanNodeIdAllocator;
+import io.trino.sql.planner.SymbolAllocator;
+import io.trino.sql.planner.TypeProvider;
+import io.trino.sql.planner.plan.ExchangeNode;
+import io.trino.sql.planner.plan.JoinNode;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.SimplePlanRewriter;
+import io.trino.sql.planner.plan.TableScanNode;
+import io.trino.sql.planner.plan.UnnestNode;
+import io.trino.sql.planner.plan.ValuesNode;
+
+import java.util.List;
+import java.util.OptionalInt;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.SystemSessionProperties.getHashPartitionCount;
+import static io.trino.SystemSessionProperties.getMinimumInputRowCountPerTask;
+import static io.trino.sql.planner.plan.SimplePlanRewriter.rewriteWith;
+import static java.lang.Double.isNaN;
+import static java.lang.Math.max;
+import static java.util.Objects.requireNonNull;
+
+public class DetermineHashPartitionCount
+        implements PlanOptimizer
+{
+    private static final Logger log = Logger.get(DetermineHashPartitionCount.class);
+
+    private final StatsCalculator statsCalculator;
+
+    public DetermineHashPartitionCount(StatsCalculator statsCalculator)
+    {
+        this.statsCalculator = requireNonNull(statsCalculator, "statsCalculator is null");
+    }
+
+    @Override
+    public PlanNode optimize(
+            PlanNode plan,
+            Session session,
+            TypeProvider types,
+            SymbolAllocator symbolAllocator,
+            PlanNodeIdAllocator idAllocator,
+            WarningCollector warningCollector,
+            TableStatsProvider tableStatsProvider)
+    {
+        requireNonNull(plan, "plan is null");
+        requireNonNull(session, "session is null");
+        requireNonNull(types, "types is null");
+        requireNonNull(tableStatsProvider, "tableStatsProvider is null");
+
+        try {
+            OptionalInt hashPartitionCount = determineHashPartitionCount(plan, session, types, tableStatsProvider);
+            if (hashPartitionCount.isPresent()) {
+                return rewriteWith(new Rewriter(hashPartitionCount.getAsInt()), plan);
+            }
+        }
+        catch (RuntimeException e) {
```

Comment:
`CachingStatsProvider` should already catch exception and ignore them.

BTW: we probably should fix underlying issues hence could you create Trino issue for stats failures that you've seen

Reply:
We can't remove this simply since TestingTrinoServer explicitly disables `ignore-stats-calculator-failures` and this rule is very generic and applicable to all kinds of queries. Therefore, many tests are failing.

Reply:
https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/server/testing/TestingTrinoServer.java#L248

Reply:
so these tests are supposed to fail but with a different error that comes at execution time. However, with this rule enabled they are failing at getTableStats